theia-ide/theia
# view containers integration

New container contributed by the plugin system:
<img width="1792" alt="Screen Shot 2019-07-09 at 16 53 28" src="https://user-images.githubusercontent.com/3082655/60899049-1ab7d880-a26a-11e9-85da-e9faf84ec1d1.png">

Views contributed to the explorer container by the plugin system:
<img width="1792" alt="Screen Shot 2019-07-11 at 16 23 30" src="https://user-images.githubusercontent.com/3082655/61058878-541e4e80-a3f8-11e9-9d06-dc13174ca1b0.png">

#### What it does

- [x] close https://github.com/theia-ide/theia/issues/5058 - complete integration of view containers in the plugin system and shell
- [x] https://github.com/theia-ide/theia/issues/4199 - support `onView` activation event 
- [x] register views after all view containers are registered
- [x] fix https://github.com/theia-ide/theia/issues/4174 - support view/title menu contribution
- [x] support more view/title menu contribution
- [x] pre-defined views
  - [x] debug
  - [x] scm
  - [x] explorer
    - [x] a CQ for copied icon - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20431
- [x] a command to toggle view container in the view menu and the command palette 
- [x] commands to toggle view containers contributed by plugin system
- [x] `Open View` command
- [x] quick open `view` prefix
- [x] implemented TreeItem2.label proposed API for sample extension
3ea5c9ddbebd8ec68e3b821f9c39c3ec785fde97/src/vs/vscode.proposed.d.ts#L1436-L1477
  - [x] a CQ for copied code - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20431
- [x] fix https://github.com/theia-ide/theia/issues/4637 - single-view view-container mode
  - [x] view caption should be inlined into the view container caption
  - [x] view toolbar items should be inlined into the view container toolbar
    - [x] scm
    - [x] explorer
    - [x] plugin views
  - [x] toolbar context menu should allow to hide/show view parts
- [x] show view container part's toolbar only on hover
- [x] plugin tree views: not enough space between icon and text if the width is small
- [x] restoration layout of view container on reload
  - [x] debug - no stateful
  - [x] scm - input message is preserved
  - [x] explorer
    - [x] fix https://github.com/theia-ide/theia/issues/1781 - restore internal state of the navigator
  - [x] plugin views
    - [x] remove stale widget, i.e. which are not registered anymore
    - [x] move widgets to new view containers if package.json was changed
    - [x] restore internal state of plugin tree widgets
- [x] view container activation
  -  [x] should restore active part on restore of shell layout
  -  [x] should restore previous active part on activation, or activate the first expanded or visible part, or receive focus itself
-  [x] shell: handle activation and reveal requests recursively
- [x] fix UI tests
- [x] shell layout versioning and backward-compatibility
- [x] integrate https://github.com/theia-ide/theia/pull/5718

#### How to test

- install https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample from sources under `plugins folder
- run Theia and check that all views are added
- modify package.json to move views to other view containers (scm, debug, test) and test that views are added there as well

What to pay attention:
- keyboard navigation: one can navigate between view parts with keyboard plus collapse and expand them, try to drag and drop as well
- activation (focus): a proper part is focused
- restoration of layout
- explorer in multi-root workspace

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#reviewing)
